### PR TITLE
[WIP] Separately track the vote counts since the previous debate

### PR DESCRIPTION
**Pull request by jbradberry**
OP-173

Still to do:
- [x] add the entries to the form for sorting by `current_votes`
- [x] write the tests

### CODE REVIEW

File: opendebates/context_processors.py [hunk 1]
```diff
@@ -63,6 +63,7 @@ def _get_categories():
         'ALLOW_SORTING_BY_VOTES': mode.allow_sorting_by_votes,
         'ALLOW_VOTING_AND_SUBMITTING_QUESTIONS': mode.allow_voting_and_submitting_questions,
         'DEBATE_TIME': mode.debate_time,
+        'PREVIOUS_DEBATE_TIME': mode.previous_debate_time,
```
Comment (tobiasmcnulty) [inline opendebates/context_processors.py]: How about just using `mode` in the template (which is already available), with the intent of eventually deprecating all the extra variables?

File: opendebates/models.py [hunk 2]
```diff
@@ -135,6 +140,7 @@ def user_display_name(self):
 
     votes = models.IntegerField(default=0, db_index=True)
     local_votes = models.IntegerField(default=0, db_index=True)
+    current_votes = models.IntegerField(default=0, db_index=True)
```
Comment (tobiasmcnulty) [inline opendebates/models.py]: At some point (probably not now) we should check this table to see if all these indexes help or hurt performance... CC @ejucovy

File: opendebates/moderator_views.py [hunk 3]
```diff
@@ -60,6 +60,13 @@ def merge(request):
         votes_to_merge = Vote.objects.filter(submission=to_remove).exclude(
             voter__in=votes_already_cast)
 
+        previous_debate_time = get_previous_debate_time()
+        if previous_debate_time:
+            current_votes_to_merge = votes_to_merge.filter(
+                created_at__gt=previous_debate_time).count()
+        else:
+            current_votes_to_merge = 0
```
Comment (tobiasmcnulty) [inline opendebates/moderator_views.py]: This seems contrary to the behavior of `current_votes`, which is incremented even if `previous_debate_time is None` (which as noted elsewhere I'm not sure is needed).

File: opendebates/templates/opendebates/list_ideas.html [hunk 4]
```diff
@@ -81,6 +81,11 @@
           <option value="+votes" {% if sort == "+votes" %}selected{% endif %}>
             {% blocktrans %}Fewest Votes{% endblocktrans %}
           </option>
+          {% if PREVIOUS_DEBATE_TIME %}
```
Comment (tobiasmcnulty) [inline opendebates/templates/opendebates/list_ideas.html:90]: `mode.previous_debate_time`

File: opendebates/utils.py [hunk 5]
```diff
@@ -131,6 +131,8 @@ def sort_list(citations_only, sort, ideas):
         ideas = ideas.order_by("votes")
     elif sort == "-local_votes":
         ideas = ideas.order_by("-local_votes")
+    elif sort == "-current_votes":
+        ideas = ideas.order_by("-current_votes")[:30]
```
Comment (tobiasmcnulty) [inline opendebates/utils.py]: I didn't think this was going to be limited to 30 in the code -- is this correct per the client?

File: opendebates/views.py [hunk 6]
```diff
@@ -239,10 +239,15 @@ def vote(request, id):
 
         )
     )
+    previous_debate_time = get_previous_debate_time()
     if created:
         # update the DB with the real tally
         Submission.objects.filter(id=id).update(
             votes=F('votes')+1,
+            current_votes=F('current_votes')+(
+                1 if previous_debate_time is None or vote.created_at > previous_debate_time
```
Comment (tobiasmcnulty) [inline opendebates/views.py:255]: Not sure there's a need to increment this at all unless `previous_debate_time` is set and in the past; i.e., should the first part of that be `if previous_debate_time is not None and ...`?

File: opendebates/views.py [hunk 7]
```diff
@@ -306,17 +311,21 @@ def questions(request):
...
         headline=form_data['headline'],
         followup=form_data['question'],
         idea=(u'%s %s' % (form_data['headline'], form_data['question'])).strip(),
         citation=form_data['citation'],
-        created_at=timezone.now(),
+        created_at=created_at,
         ip_address=get_ip_address_from_request(request),
         approved=True,
         votes=1,
+        current_votes=(1 if previous_debate_time is None or created_at > previous_debate_time
```
Comment (tobiasmcnulty) [inline opendebates/views.py:334]: See above.

File: opendebates/views.py [hunk 8]
```diff
@@ -239,10 +239,15 @@ def vote(request, id):
 
         )
     )
+    previous_debate_time = get_previous_debate_time()
```
Comment (tobiasmcnulty) [inline opendebates/views.py:249]: Just noting that this adds another cache lookup to this frequently-used view.

File: opendebates/tests/test_vote.py [hunk 9]
```diff
@@ -294,6 +306,53 @@ def test_vote_user_disabled_captcha(self):
...
+        rsp = self.client.post(self.submission_url, data=data,
+                               HTTP_X_REQUESTED_WITH='XMLHttpRequest')
+        self.assertEqual(200, rsp.status_code)
+        json_rsp = json.loads(rsp.content)
+        # vote is incremented in JSON response
+        self.assertEqual(self.votes + 1, json_rsp['tally'])
+        refetched_sub = Submission.objects.get(pk=self.submission.pk)
+        # ... and in DB
+        self.assertEqual(self.votes + 1, refetched_sub.votes)
+        self.assertEqual(self.current_votes, refetched_sub.current_votes)
```
Comment (tobiasmcnulty) [inline opendebates/tests/test_vote.py]: It took me a moment to find the actual thing being tested in these tests -- perhaps a comment above this line (and the corresponding line in the previous test) would be in order?

Optionally, could any of the other boilerplate be removed, since it appears to be well-tested elsewhere?

File: opendebates/admin.py [hunk 10]
```diff
@@ -103,3 +104,8 @@ class SiteModeAdmin(ModelAdmin):
     _missed_fields = set(_fields) - _used_fields
     if _missed_fields:
         fieldsets.append(('Other', {'fields': sorted(list(_missed_fields))}))
+
+    def reset_current_votes(self, request, queryset):
+        # Note, currently all submissions are under the one and only SiteMode.
+        models.Submission.objects.update(current_votes=0)
```
Comment (tobiasmcnulty) [inline opendebates/admin.py]: I have no objection to doing it this way (in fact I like that it's more explicit), but the way I explained this to Lilia previously it happened as part of updating `previous_debate_time`. You or Katie will want to update her.

This also makes me wonder if we could support updating `previous_debate_time` retroactively, but after a few minutes of digging I couldn't find a way to update `current_votes` atomically in Django (based on `previous_debate_time`) without manually writing the subquery.

**Review by tobiasmcnulty — COMMENTED** [review]
I left a few comments, mostly on things where I thought the business logic was slightly different. I didn't have time to run it locally so it'd be nice to get a full review from @dpoirier, @vkurup, or @ejucovy, if possible.

**Comment by tobiasmcnulty** [comment]
I did test it so far as to see that it won't be possible to run the old code on the migrated DB. So, unless we push this out tonight and tweak it later, downtime will be required to deploy this.

```
Environment:


Request Method: POST
Request URL: http://localhost:8000/questions/

Django Version: 1.8.12
Python Version: 2.7.10
Installed Applications:
['django.contrib.auth',
 'django.contrib.contenttypes',
 'django.contrib.sessions',
 'django.contrib.messages',
 'django.contrib.staticfiles',
 'django.contrib.sites',
 'django.contrib.flatpages',
 'pipeline',
 'djangobower',
 'dbbackup',
 'nocaptcha_recaptcha',
 'djcelery',
 'opendebates',
 'opendebates_emails',
 'djorm_pgfulltext',
 'endless_pagination',
 'bootstrapform',
 'registration',
 'django.contrib.admin']
Installed Middleware:
('django.contrib.sessions.middleware.SessionMiddleware',
 'django.middleware.locale.LocaleMiddleware',
 'django.middleware.common.CommonMiddleware',
 'django.middleware.csrf.CsrfViewMiddleware',
 'django.contrib.flatpages.middleware.FlatpageFallbackMiddleware',
 'django.contrib.auth.middleware.AuthenticationMiddleware',
 'django.contrib.auth.middleware.SessionAuthenticationMiddleware',
 'django.contrib.messages.middleware.MessageMiddleware',
 'django.middleware.clickjacking.XFrameOptionsMiddleware',
 'django.middleware.security.SecurityMiddleware',
 'opendebates.router.DBRoutingMiddleware')


Traceback:
File "/Users/tobias/.virtualenvs/opendebates/lib/python2.7/site-packages/django/core/handlers/base.py" in get_response
  132.                     response = wrapped_callback(request, *callback_args, **callback_kwargs)
File "/Users/tobias/.virtualenvs/opendebates/lib/python2.7/site-packages/djangohelpers/lib.py" in rendered_func
  22.             response = func(request, *args, **kwargs)
File "/Users/tobias/.virtualenvs/opendebates/lib/python2.7/site-packages/djangohelpers/lib.py" in inner
  48.             return func(request, *args, **kwargs)
File "/Users/tobias/caktus/django-opendebates/opendebates/views.py" in questions
  321.         source=request.COOKIES.get('opendebates.source'),
File "/Users/tobias/.virtualenvs/opendebates/lib/python2.7/site-packages/django/db/models/manager.py" in manager_method
  127.                 return getattr(self.get_queryset(), name)(*args, **kwargs)
File "/Users/tobias/.virtualenvs/opendebates/lib/python2.7/site-packages/django/db/models/query.py" in create
  348.         obj.save(force_insert=True, using=self.db)
File "/Users/tobias/.virtualenvs/opendebates/lib/python2.7/site-packages/django/db/models/base.py" in save
  734.                        force_update=force_update, update_fields=update_fields)
File "/Users/tobias/.virtualenvs/opendebates/lib/python2.7/site-packages/django/db/models/base.py" in save_base
  762.             updated = self._save_table(raw, cls, force_insert, force_update, using, update_fields)
File "/Users/tobias/.virtualenvs/opendebates/lib/python2.7/site-packages/django/db/models/base.py" in _save_table
  846.             result = self._do_insert(cls._base_manager, using, fields, update_pk, raw)
File "/Users/tobias/.virtualenvs/opendebates/lib/python2.7/site-packages/django/db/models/base.py" in _do_insert
  885.                                using=using, raw=raw)
File "/Users/tobias/.virtualenvs/opendebates/lib/python2.7/site-packages/django/db/models/manager.py" in manager_method
  127.                 return getattr(self.get_queryset(), name)(*args, **kwargs)
File "/Users/tobias/.virtualenvs/opendebates/lib/python2.7/site-packages/django/db/models/query.py" in _insert
  920.         return query.get_compiler(using=using).execute_sql(return_id)
File "/Users/tobias/.virtualenvs/opendebates/lib/python2.7/site-packages/django/db/models/sql/compiler.py" in execute_sql
  974.                 cursor.execute(sql, params)
File "/Users/tobias/.virtualenvs/opendebates/lib/python2.7/site-packages/django/db/backends/utils.py" in execute
  79.             return super(CursorDebugWrapper, self).execute(sql, params)
File "/Users/tobias/.virtualenvs/opendebates/lib/python2.7/site-packages/django/db/backends/utils.py" in execute
  64.                 return self.cursor.execute(sql, params)
File "/Users/tobias/.virtualenvs/opendebates/lib/python2.7/site-packages/django/db/utils.py" in __exit__
  98.                 six.reraise(dj_exc_type, dj_exc_value, traceback)
File "/Users/tobias/.virtualenvs/opendebates/lib/python2.7/site-packages/django/db/backends/utils.py" in execute
  64.                 return self.cursor.execute(sql, params)

Exception Type: IntegrityError at /questions/
Exception Value: null value in column "current_votes" violates not-null constraint
DETAIL:  Failing row contains (9, test test, test, test, , f, 2016-09-26 22:50:37.805745+00, 127.0.0.1, f, t, f, 1, 0, 0, 0, , null, 1, null, 2, null, f, 0, null).
```

**Comment by jbradberry** [comment]
@ejucovy I'd also like to get your thoughts on this, if you have time.

File: opendebates/models.py [hunk 11]
```diff
@@ -41,6 +41,11 @@ class SiteMode(CachingMixin, models.Model):
         help_text="Enter time that debate starts in timezone %s" % settings.TIME_ZONE,
     )
     debate_state = models.CharField(max_length=5, null=True, blank=True)
+    previous_debate_time = models.DateTimeField(
+        null=True, blank=True,
+        help_text=("Enter time that the previous debate occurred in timezone %s to"
+                   " reset current vote counts" % settings.TIME_ZONE)
```
Comment (ejucovy) [inline opendebates/models.py]: "to reset current vote counts" is a bit misleading if the admin action remains separate - maybe just "to enable Votes Since Previous Debate sort option"?

File: opendebates/templates/opendebates/list_ideas.html [hunk 12]
```diff
@@ -81,6 +81,11 @@
...
+          <option value="-current_votes" {% if sort == "-current_votes" %}selected{% endif %}>
+            {% blocktrans %}Top 30 Since Last Debate{% endblocktrans %}
```
Comment (ejucovy) [inline opendebates/templates/opendebates/list_ideas.html]: Per Tobias' comment elsewhere, I _think_ we're not supposed to actually limit this query to 30.  So maybe just "Most Votes Since Last Debate" here?  But this should all be confirmed with Lilia.

File: opendebates/tests/test_submission.py [hunk 13]
```diff
@@ -97,6 +99,38 @@ def test_post_submission(self):
...
         self.assertEqual(submission.votes, 1)
+        self.assertEqual(submission.current_votes, 1)
+        # but Vote objects are also created
+        votes = Vote.objects.filter(voter=self.voter)
+        self.assertEqual(1, len(votes))
+
+    def test_post_submission_after_previous_debate(self):
+        mode, _ = SiteMode.objects.get_or_create()
+        mode.previous_debate_time = timezone.make_aware(
+            datetime.datetime(2016, 1, 1, 12)
```
Comment (ejucovy) [inline opendebates/tests/test_submission.py]: Could this be `timezone.now() - datetime.timedelta(something)` instead?  I have a weird aversion to introducing a dependency on the code having being written after Jan 2016 (and being run on a machine whose clock isn't completely wrong).  Would also make the intention more explicit on a quick skim.

File: opendebates/tests/test_views.py [hunk 14]
```diff
@@ -23,3 +26,21 @@ def test_list_ideas_uses_site_mode_from_context(self, gsm_mock):
...
+
+    def test_top_30_visible_if_previous_debate(self):
+        mode, _ = SiteMode.objects.get_or_create()
+        mode.previous_debate_time = timezone.make_aware(
+            datetime.datetime(2016, 1, 1, 12)
+        )
+        mode.save()
+
+        response = self.client.get(self.url)
+        self.assertContains(response, "Top 30 Since Last Debate")
```
Comment (ejucovy) [inline opendebates/tests/test_views.py]: Hmm.  Is top 30 visible if previous debate is non-null but in the future?

I _think_ ideally it wouldn't be -- that way moderators could actually set it ahead of time and forget about it, relying on the system to start incrementing and make the sort option visible only after the right time.

File: opendebates/tests/test_vote.py [hunk 15]
```diff
@@ -294,6 +306,53 @@ def test_vote_user_disabled_captcha(self):
         refetched_sub = Submission.objects.get(pk=self.submission.pk)
         # ... and in DB
         self.assertEqual(self.votes + 1, refetched_sub.votes)
+        self.assertEqual(self.current_votes + 1, refetched_sub.current_votes)
+
+    def test_vote_after_previous_debate(self):
+        "Votes after the previous debate get tracked separately."
+        mode, _ = SiteMode.objects.get_or_create()
+        mode.previous_debate_time = timezone.make_aware(
+            datetime.datetime(2016, 1, 1, 12)
```
Comment (ejucovy) [inline opendebates/tests/test_vote.py]: Same comment as above - could we derive this from timezone.now() to be more explicit about what matters to the test?

File: opendebates/utils.py [hunk 16]
```diff
@@ -131,6 +131,8 @@ def sort_list(citations_only, sort, ideas):
         ideas = ideas.order_by("votes")
     elif sort == "-local_votes":
         ideas = ideas.order_by("-local_votes")
+    elif sort == "-current_votes":
```
Comment (ejucovy) [inline opendebates/utils.py:134]: Is it worth checking whether the sitemode state is currently allowing this sort order, to prevent people from manually constructing a query string with it when the sitemode option is unset?  (Most of time that result would just be uninteresting, so maybe not worth the extra logic.)

File: opendebates/admin.py (hunk 10 above)
Comment (ejucovy) [reply opendebates/admin.py]: I like the explicitness too, but, actually, what is the sequence that moderators would have to follow here?  I think things will have to be done in a specific order to keep the tallies accurate:
1. Set previous_debate_time to something in the future
2. Run the admin action
3. Wait for time to pass

If 1<->2 are flipped, or if you set previous_debate_time to something that isn't yet in the future, ISTM we'd end up with tallies that aren't quite accurate.

If this is correct, then I think it also supports my suggestion that the sort option should only become visible if previous_debate_time is in the past - otherwise there'd be no way to take the necessary steps without turning on the UI prematurely.

File: opendebates/context_processors.py (hunk 1 above)
Comment (ejucovy) [reply opendebates/context_processors.py]: Per my comments elsewhere I think we may actually want this to be more like `mode.previous_debate_time if mode.previous_debate_time and mode.previous_debate_time < now() else None`.

File: opendebates/views.py (hunk 6 above)
Comment (jbradberry) [reply opendebates/views.py:255]: I was aiming for having the counts be immediately ready if and when the feature gets enabled.

File: opendebates/context_processors.py (hunk 1 above)
Comment (jbradberry) [reply opendebates/context_processors.py]: Since per Ethan's comment this rightly should involve a bit of logic, I kept this line but updated in the way suggested.

File: opendebates/moderator_views.py (hunk 3 above)
Comment (jbradberry) [reply opendebates/moderator_views.py]: Updated to do the full count, since I wanted all `current_votes` to be ready and up to date if and when the feature is enabled.

File: opendebates/utils.py (hunk 5 above)
Comment (jbradberry) [reply opendebates/utils.py]: Yeah, I think that makes sense.  Dropping the 30.

File: opendebates/tests/test_vote.py (hunk 9 above)
Comment (jbradberry) [reply opendebates/tests/test_vote.py]: Ok, I've stripped the boilerplate assertions, and added a comment.

File: opendebates/admin.py (hunk 10 above)
Comment (jbradberry) [reply opendebates/admin.py]: @ejucovy yes, you are correct, 1 & 2 would have to be executed in the correct order.  I've updated the other locations to do your suggested check for the `previous_debate_time` still being in the future.

File: opendebates/models.py (hunk 11 above)
Comment (jbradberry) [reply opendebates/models.py]: Updated.

File: opendebates/templates/opendebates/list_ideas.html (hunk 12 above)
Comment (jbradberry) [reply opendebates/templates/opendebates/list_ideas.html]: Agreed.  I've updated both the text and the queryset.

File: opendebates/tests/test_submission.py (hunk 13 above)
Comment (jbradberry) [reply opendebates/tests/test_submission.py]: Good call.  I've updated all of the tests where I did this to use `timezone.now() - datetime.timedelta(days=7)`.

File: opendebates/tests/test_views.py (hunk 14 above)
Comment (jbradberry) [reply opendebates/tests/test_views.py]: I've added this test.

File: opendebates/templates/opendebates/list_ideas.html (hunk 4 above)
Comment (kmtracey) [reply opendebates/templates/opendebates/list_ideas.html:90]: PREVIOUS_DEBATE_TIME is set by context processor and ensures the mode.previous_debate_time value is actually in the past...is that why this comment didn't result in a change?

Comment (jbradberry) [reply opendebates/templates/opendebates/list_ideas.html:90]: Yes.  I decided that since, as Ethan pointed out, the template needed a bit of logic more complicated than just testing for non-`None`, that it should be rolled into that context processor variable.
